File: bin/string_gen.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <regex>
#include <vector>
#include <unordered_map>
#include <string>
#include <nlohmann/json.hpp>  // Include the nlohmann/json.hpp library

using json = nlohmann::json;

// Function to convert the hexadecimal strings to bytes
std::vector<uint8_t> convertFormat(const std::string& hexString) {
    std::vector<uint8_t> result;
    size_t i = 0;
    while (i < hexString.length()) {
        if (hexString[i] == '|') {
            ++i;
            size_t start = i;
            while (i < hexString.length() && hexString[i] != '|') {
                ++i;
            }
            std::string part = hexString.substr(start, i - start);
            bool isHex = std::all_of(part.begin(), part.end(), [](char c) {
                return std::isxdigit(c) || c == ' ';
            });

            if (isHex) {
                std::istringstream hexStream(part);
                std::string hexValue;
                while (hexStream >> hexValue) {
                    result.push_back(static_cast<uint8_t>(std::stoi(hexValue, nullptr, 16)));
                }
            } else {
                result.insert(result.end(), part.begin(), part.end());
            }
        } else {
            result.push_back(static_cast<uint8_t>(hexString[i]));
        }
        ++i;
    }
    return result;
}

// Function to write PCRE data to a file
void writePCREToFile(const std::unordered_map<int, json>& pcreDict, const std::string& filename) {
    std::ofstream outFile(filename);
    if (!outFile) {
        std::cerr << "Error opening file: " << filename << std::endl;
        return;
    }

    for (const auto& [key, value] : pcreDict) {
        outFile << key << ":" << value["string"].get<std::string>() << "\n";
    }
}

// Function to parse rules from the file and extract content
void parseRules(const std::string& file) {
    std::ifstream infile(file);
    if (!infile) {
        std::cerr << "Error opening rule file: " << file << std::endl;
        return;
    }

    std::unordered_map<int, json> stringTable, ruleTable;
    std::vector<std::vector<uint8_t>> strings;
    std::string line;
    
    // Updated regex pattern to match double quotes properly
    std::regex contentPattern(R"(\\\"([^\\\"]*)\\\")");

    int ruleId = 0, count = 0;
    while (std::getline(infile, line)) {
        std::smatch match;
        if (std::regex_search(line, match, contentPattern)) {
            ruleId++;
            ruleTable[ruleId] = { {"sid", ruleId}, {"str_id", json::array()} };

            for (const auto& value : match) {
                std::string patt = value.str();
                std::vector<uint8_t> stringData = convertFormat(patt);

                if (stringData.size() > 1) {
                    count++;
                    strings.push_back(stringData);
                    stringTable[count] = { {"sid", ruleId}, {"string", std::string(stringData.begin(), stringData.end())}, {"rid", ruleId} };
                    ruleTable[ruleId]["str_id"].push_back(count);
                }
            }
        }
    }

    // Convert the unordered_map to a json object for dumping
    json ruleTableJson(ruleTable);
    json stringTableJson(stringTable);

    // Write output to files
    std::ofstream ruleFile("rule_table_cpp.json");
    ruleFile << ruleTableJson.dump(4);

    std::ofstream stringFile("string_table_cpp.json");
    stringFile << stringTableJson.dump(4);

    writePCREToFile(stringTable, "literals_cpp.txt");
}

int main() {
    std::string file = "snort3-community.rules";  // File name
    parseRules(file);
    return 0;
}

File: bin/string_gen.py
from idstools import rule
import json
import re

# file = "eternalblue_rule.rules"
file = "snort3-community.rules"

def convert_format(hex_string):
    result = bytearray()  # Using bytearray to efficiently build the result byte string
    i = 0
    while i < len(hex_string):
        if hex_string[i] == '|':
            i += 1
            start = i
            while i < len(hex_string) and hex_string[i] != '|':
                i += 1
            # Process the enclosed segment
            part = hex_string[start:i]
            if all(c in '0123456789ABCDEFabcdef ' for c in part):
                # Convert hex string to bytes
                hex_values = part.split()
                for hex_value in hex_values:
                    result.extend(bytes.fromhex(hex_value))
            else:
                # Handle non-hex segments as ASCII
                result.extend(part.encode('utf-8'))
        else:
            # Handle non-enclosed parts
            result.append(ord(hex_string[i]))
        i += 1
    return bytes(result)

def write_pcre_to_file(pcre_dict, filename):
    with open(filename, 'w') as file:
        for key, value in pcre_dict.items():
            pattern = value['string']#.strip('"')  # Strip quotes
            file.write(f"{key}:{pattern}\n")

strings = [] # strings extracted from rules
string_table = {}
rule_table = {}
pattern = r'"([^"]*)"'
r_id = 0
count=0

for rule in rule.parse_file(file):
    if "content" in rule:
        content_values = [option['value'] for option in rule['options'] if option['name'] == 'content']
        r_id += 1
        rule_table[r_id] = {'sid': rule.sid,
                            'str_id': []}
        for value in content_values:
            match  = re.search(pattern, value)
            patt = match.group(1)

            if len(patt) > 1:
                string = convert_format(patt)

                if match and len(string) > 1:
                    count += 1
                    strings.append(string)
                    string_table[count] = {'sid': rule.sid,
                                          'string': str(string)[2:-1],
                                          'rid': r_id}
                    rule_table[r_id]['str_id'].append(count)
with open('rule_table.json', 'w') as rule_file:
    json.dump(rule_table, rule_file, indent=4)

# Saving string_table to JSON file
with open('string_table.json', 'w') as string_file:
    json.dump(string_table, string_file, indent=4)
                    
write_pcre_to_file(string_table, 'literals.txt')
